File: ios/Releaf/Features/Notepad/NotepadCalendarBloom.swift
/*
 * NotepadCalendarBloom.swift
 *
 * The Day view's centerpiece — a 5×7 grid of small pines, one per day
 * of the current month. Empty days render as hollow tree silhouettes;
 * days with entries fill the canopy in three tinted greens (light →
 * mid → deep) keyed off the day's `DayDensity`. Today is a coral pine
 * with a coral ring.
 *
 * Tree silhouette + color ramp mirror TreesSavedHeroView.TreeGlyph and
 * the side-drawer canopy header so the three surfaces feel like the
 * same forest at different framings.
 */

import SwiftUI
import ReleafDesignSystem

private let emptyFill   = Color(red: 0xFA / 255.0, green: 0xF1 / 255.0, blue: 0xE6 / 255.0)
private let lightFill   = Color(red: 0xC0 / 255.0, green: 0xDD / 255.0, blue: 0x97 / 255.0)
private let midFill     = Color(red: 0x7A / 255.0, green: 0xA8 / 255.0, blue: 0x74 / 255.0)
private let deepFill    = Color(red: 0x3E / 255.0, green: 0x6B / 255.0, blue: 0x3B / 255.0)
private let todayCoral  = Color(red: 0xE7 / 255.0, green: 0x78 / 255.0, blue: 0x50 / 255.0)
private let todayCoralDark = Color(red: 0x99 / 255.0, green: 0x3C / 255.0, blue: 0x1D / 255.0)
private let trunkColor  = Color(red: 0x3E / 255.0, green: 0x2A / 255.0, blue: 0x18 / 255.0)
private let trunkSoft   = Color(red: 0x5F / 255.0, green: 0x4A / 255.0, blue: 0x2D / 255.0)
private let coralHalo   = Color(red: 0xFC / 255.0, green: 0xEA / 255.0, blue: 0xE0 / 255.0)
private let greenLine   = Color(red: 0x1E / 255.0, green: 0x59 / 255.0, blue: 0x43 / 255.0)
private let selectionRing = Color(red: 0x5B / 255.0, green: 0x8C / 255.0, blue: 0x52 / 255.0)
private let selectionHalo = Color(red: 0xD9 / 255.0, green: 0xED / 255.0, blue: 0xE2 / 255.0)

private func fill(for density: DayDensity) -> Color {
    switch density {
    case .empty: return emptyFill
    case .light: return lightFill
    case .mid:   return midFill
    case .deep:  return deepFill
    }
}

public struct NotepadCalendarBloom: View {
    let leadingBlanks: Int
    let days: [DayCount]
    let today: Date
    let onDayTap: (DayCount) -> Void
    let showLegend: Bool
    let showWeekdayStrip: Bool
    let selectedDate: Date?

    public init(
        leadingBlanks: Int,
        days: [DayCount],
        today: Date,
        onDayTap: @escaping (DayCount) -> Void,
        showLegend: Bool = true,
        showWeekdayStrip: Bool = true,
        selectedDate: Date? = nil
    ) {
        self.leadingBlanks = leadingBlanks
        self.days = days
        self.today = today
        self.onDayTap = onDayTap
        self.showLegend = showLegend
        self.showWeekdayStrip = showWeekdayStrip
        self.selectedDate = selectedDate
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            if showWeekdayStrip {
                NotepadCalendarWeekdayStrip()
            }

            // Tree grid — pad with leading + trailing blanks so the 7-wide layout
            // aligns with the weekday columns.
            let cells: [DayCount?] = leadingNils + days.map { Optional($0) } + trailingNils
            let rows = cells.count / 7
            VStack(spacing: 0) {
                ForEach(0..<rows, id: \.self) { row in
                    HStack(spacing: 0) {
                        ForEach(0..<7, id: \.self) { col in
                            let cell = cells[row * 7 + col]
                            let cal = Calendar.current
                            let isToday = cell.map { cal.isDate($0.date, inSameDayAs: today) } ?? false
                            // Selection wins over "today" for the ring —
                            // when today is the selected day (default on
                            // open), the green ring shows. Today's coral
                            // canopy fill still marks the date.
                            let isSelected: Bool = {
                                guard let sel = selectedDate, let c = cell else { return false }
                                return cal.isDate(c.date, inSameDayAs: sel)
                            }()
                            TreeCell(
                                day: cell,
                                isToday: isToday,
                                isSelected: isSelected,
                                onTap: { c in onDayTap(c) }
                            )
                            .frame(maxWidth: .infinity, minHeight: 36)
                        }
                    }
                }
            }

            // Legend (optional — hidden when used inside the 3-month
            // carousel where only the centered month should show its
            // legend, rendered separately by the caller).
            if showLegend {
                LegendRow()
                    .padding(.top, 4)
            }
        }
    }


    // MARK: - Padding

    private var leadingNils: [DayCount?] {
        Array(repeating: nil, count: leadingBlanks)
    }

    private var trailingNils: [DayCount?] {
        let total = leadingBlanks + days.count
        let pad = (7 - total % 7) % 7
        return Array(repeating: nil, count: pad)
    }

    private static let weekdayInitials = ["S", "M", "T", "W", "T", "F", "S"]
}

// MARK: - Tree cell

private struct TreeCell: View {
    let day: DayCount?
    let isToday: Bool
    let isSelected: Bool
    let onTap: (DayCount) -> Void

    var body: some View {
        Group {
            if let day {
                Button(action: { onTap(day) }) {
                    Canvas { ctx, size in
                        let cx = size.width / 2
                        let cy = size.height / 2 - 2
                        let r = min(size.width, size.height) / 2
                        let halo = Path(ellipseIn: CGRect(x: cx - r, y: cy - r, width: r * 2, height: r * 2))

                        if isSelected {
                            // Selection takes ring priority — works for
                            // today (which is selected by default on
                            // open) and any past/future day the user
                            // taps. Today's coral canopy fill still
                            // distinguishes the date even when its ring
                            // goes green.
                            ctx.fill(halo, with: .color(selectionHalo.opacity(0.55)))
                            ctx.stroke(halo, with: .color(selectionRing.opacity(0.85)), lineWidth: 1.5)
                        } else if isToday {
                            // Today, but not currently the selected day —
                            // user has tapped a past/future day. Coral
                            // ring stays as the "today marker" until
                            // they tap back on today.
                            ctx.fill(halo, with: .color(coralHalo.opacity(0.6)))
                            ctx.stroke(halo, with: .color(todayCoral.opacity(0.85)), lineWidth: 1.5)
                        }

                        drawTree(in: ctx, center: CGPoint(x: cx, y: cy), density: day.density, isToday: isToday)
                    }
                    .frame(width: 28, height: 32)
                }
                .buttonStyle(.plain)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
            } else {
                Color.clear.frame(maxWidth: .infinity, maxHeight: .infinity)
            }
        }
    }
}

private func drawTree(in ctx: GraphicsContext, center: CGPoint, density: DayDensity, isToday: Bool) {
    let cx = center.x
    let cy = center.y
    var canopy = Path()
    canopy.move(to: CGPoint(x: cx, y: cy - 9))
    canopy.addLine(to: CGPoint(x: cx - 7, y: cy + 5))
    canopy.addLine(to: CGPoint(x: cx + 7, y: cy + 5))
    canopy.closeSubpath()

    let canopyFill: Color = isToday ? todayCoral : fill(for: density)
    let isHollow = !isToday && density == .empty

    ctx.fill(canopy, with: .color(canopyFill))
    if isHollow {
        ctx.stroke(canopy, with: .color(greenLine.opacity(0.22)), lineWidth: 0.6)
    }

    // Trunk
    let trunkFill: Color = {
        if isToday { return todayCoralDark }
        if isHollow { return trunkSoft.opacity(0.4) }
        return trunkColor
    }()
    let trunkRect = CGRect(x: cx - 1.5, y: cy + 5, width: 3, height: 4)
    ctx.fill(Path(trunkRect), with: .color(trunkFill))
}

// MARK: - Legend

/// Public legend strip — callers can render it once below the
/// 3-month carousel where individual calendars hide their own legend.
///
/// When `onTodayTap` is provided, the right-aligned "today" badge
/// becomes a tap target so the host can snap the carousel back to
/// today's month and reselect today after the user has swiped or
/// tapped a different day.
public struct NotepadCalendarLegend: View {
    private let onTodayTap: (() -> Void)?

    public init(onTodayTap: (() -> Void)? = nil) {
        self.onTodayTap = onTodayTap
    }

    public var body: some View { LegendRow(onTodayTap: onTodayTap) }
}

/// Public weekday header strip — extracted so the swipeable carousel
/// can render it once above the pager, matching Android. Without
/// extraction, the prev / next page's own strips bleed into the
/// centered page during transitions.
public struct NotepadCalendarWeekdayStrip: View {
    public init() {}
    public var body: some View {
        HStack(spacing: 0) {
            ForEach(["S", "M", "T", "W", "T", "F", "S"], id: \.self) { letter in
                Text(letter)
                    .font(AppText.tag)
                    .foregroundStyle(AppColors.textSecondary)
                    .frame(maxWidth: .infinity)
            }
        }
    }
}

private struct LegendRow: View {
    let onTodayTap: (() -> Void)?

    init(onTodayTap: (() -> Void)? = nil) {
        self.onTodayTap = onTodayTap
    }

    var body: some View {
        HStack(spacing: 8) {
            Text("none")
                .font(AppText.tag)
                .foregroundStyle(AppColors.textSecondary)

            LegendTree(density: .empty)
            LegendTree(density: .light)
            LegendTree(density: .mid)
            LegendTree(density: .deep)

            Text("3+ captures")
                .font(AppText.tag)
                .foregroundStyle(AppColors.textSecondary)

            Spacer(minLength: 0)

            // Today badge — tappable when [onTodayTap] is provided. The
            // host pairs the click with snapping the carousel back to
            // today's month and reselecting today's tree.
            todayBadge
        }
    }

    @ViewBuilder
    private var todayBadge: some View {
        let badge = HStack(spacing: 6) {
            ZStack {
                Circle()
                    .stroke(todayCoral, lineWidth: 1)
                    .frame(width: 20, height: 20)
                Canvas { ctx, size in
                    drawTree(
                        in: ctx,
                        center: CGPoint(x: size.width / 2, y: size.height / 2 - 2),
                        density: .deep,
                        isToday: true
                    )
                }
                .frame(width: 14, height: 18)
            }

            Text("today")
                .font(AppText.tag)
                .foregroundStyle(onTodayTap != nil ? todayCoralDark : AppColors.textSecondary)
        }
        .padding(.horizontal, 4)
        .padding(.vertical, 2)
        .contentShape(Capsule())

        if let onTodayTap {
            Button(action: onTodayTap) { badge }
                .buttonStyle(.plain)
        } else {
            badge
        }
    }
}

private struct LegendTree: View {
    let density: DayDensity

    var body: some View {
        Canvas { ctx, size in
            drawTree(
                in: ctx,
                center: CGPoint(x: size.width / 2, y: size.height / 2 - 2),
                density: density,
                isToday: false
            )
        }
        .frame(width: 14, height: 18)
    }
}

File: android/app/src/main/java/app/releaf/mobile/features/notepad/NotepadCalendarBloom.kt
/*
 * NotepadCalendarBloom.kt
 *
 * The Day view's centerpiece — a 5×7 grid of small pines, one per day
 * of the current month. Empty days render as hollow tree silhouettes;
 * days with entries fill the canopy in three tinted greens (light →
 * mid → deep) keyed off the day's [DayDensity]. Today is a coral pine
 * with a coral ring so it lifts off the page from across the screen.
 *
 * The grid is purely visual: tap callbacks are wired so a future
 * iteration can let the user open any day's entry; for now only today
 * routes through.
 *
 * The tree silhouette + color ramp mirror TreesSavedHero.TreeGlyph and
 * the side-drawer canopy header so the three surfaces feel like the
 * same forest at different framings.
 */

package app.releaf.mobile.features.notepad

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.releaf.mobile.ui.theme.AppColors
import app.releaf.mobile.ui.theme.AppSpacing
import app.releaf.mobile.ui.theme.AppTypography
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.TextStyle
import java.util.Locale

// ---- Tree palette ----

private val EmptyFill   = Color(0xFFFAF1E6)
private val LightFill   = Color(0xFFC0DD97)
private val MidFill     = Color(0xFF7AA874)
private val DeepFill    = Color(0xFF3E6B3B)
private val TodayCoral  = Color(0xFFE77850)
private val TodayCoralDark = Color(0xFF993C1D)
private val TrunkColor  = Color(0xFF3E2A18)
private val TrunkSoft   = Color(0xFF5F4A2D)
private val GroundLine  = Color(0xFFB8956A)
private val GreenLine   = Color(0xFF1E5943)
/** Ring drawn around the user-tapped day in the calendar — same hue
 *  family as the active tree canopies so it reads as "garden focus,"
 *  while staying clearly distinct from today's coral pin. */
private val SelectionRing = Color(0xFF5B8C52)

private fun fillFor(density: DayDensity): Color = when (density) {
    DayDensity.Empty -> EmptyFill
    DayDensity.Light -> LightFill
    DayDensity.Mid   -> MidFill
    DayDensity.Deep  -> DeepFill
}

@Composable
fun NotepadCalendarBloom(
    month: YearMonth,
    days: List<DayCount>,
    today: LocalDate,
    onDayTap: (DayCount) -> Unit,
    modifier: Modifier = Modifier,
    showLegend: Boolean = true,
    showWeekdayStrip: Boolean = true,
    selectedDate: LocalDate? = null,
) {
    Column(modifier = modifier.fillMaxWidth()) {
        if (showWeekdayStrip) {
            NotepadCalendarWeekdayStrip()
            Spacer(Modifier.height(8.dp))
        }

        // 5×7 tree grid, drawn into one Canvas for tight layout
        val cols = 7
        val rowHeight = 36.dp
        val rows = totalRows(month)

        // Pad days with leading + trailing nulls so the grid lines up
        // against the weekday columns of the month.
        val leading = month.atDay(1).dayOfWeek.value % 7  // Sunday-based
        val totalCells = rows * cols
        val cells: List<DayCount?> = buildList {
            repeat(leading) { add(null) }
            addAll(days)
            while (size < totalCells) add(null)
        }

        Column(verticalArrangement = Arrangement.spacedBy(0.dp)) {
            for (row in 0 until rows) {
                Row(modifier = Modifier.fillMaxWidth().height(rowHeight)) {
                    for (col in 0 until cols) {
                        val cell = cells[row * cols + col]
                        val isToday = cell?.date == today
                        TreeCell(
                            day = cell,
                            isToday = isToday,
                            // Selection wins over "today" for the ring —
                            // when today is the selected day (default on
                            // open), the green ring shows. Today's coral
                            // canopy fill still marks the date.
                            isSelected = cell?.date == selectedDate,
                            onTap = { c -> onDayTap(c) },
                            modifier = Modifier.weight(1f).fillMaxWidth(),
                        )
                    }
                }
            }
        }

        if (showLegend) {
            Spacer(Modifier.height(AppSpacing.s3))
            NotepadCalendarLegend()
        }
    }
}

/** Density legend strip — extracted so callers can render it once
 *  even when the calendar is shown in a 3-month carousel.
 *
 *  When [onTodayTap] is provided, the right-aligned "today" badge
 *  becomes a tap target that lets the user jump the calendar back to
 *  today's month + reselect today after they've swiped or tapped a
 *  different day. */
@Composable
fun NotepadCalendarLegend(
    modifier: Modifier = Modifier,
    onTodayTap: (() -> Unit)? = null,
) {
    LegendRow(modifier = modifier, onTodayTap = onTodayTap)
}

/** Weekday header strip (S M T W T F S). Extracted so the swipeable
 *  carousel can render it once above the pager — otherwise the prev /
 *  next month's own weekday strips bleed into the centered page's
 *  edges and look like a 9–11-letter mash-up. */
@Composable
fun NotepadCalendarWeekdayStrip(modifier: Modifier = Modifier) {
    Row(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceEvenly,
    ) {
        // Sun → Sat
        for (i in 0..6) {
            val dow = java.time.DayOfWeek.SUNDAY.plus(i.toLong())
            Text(
                text = dow.getDisplayName(TextStyle.NARROW, Locale.getDefault()),
                style = AppTypography.Tag,
                color = AppColors.TextSecondary,
                modifier = Modifier.width(36.dp),
                textAlign = androidx.compose.ui.text.style.TextAlign.Center,
            )
        }
    }
}

private fun totalRows(month: YearMonth): Int {
    val leading = month.atDay(1).dayOfWeek.value % 7
    val total = leading + month.lengthOfMonth()
    return (total + 6) / 7
}

@Composable
private fun TreeCell(
    day: DayCount?,
    isToday: Boolean,
    isSelected: Boolean,
    onTap: (DayCount) -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .then(
                if (day != null) Modifier.clickable { onTap(day) } else Modifier
            ),
        contentAlignment = Alignment.Center,
    ) {
        if (day != null) {
            Canvas(modifier = Modifier.size(28.dp, 32.dp)) {
                val cx = size.width / 2f
                val cyTreeMid = size.height / 2f - 2.dp.toPx()
                if (isSelected) {
                    // Selection takes ring priority — works for today
                    // (which is selected by default on open) and any
                    // past/future day the user taps. Today's coral
                    // tree fill still distinguishes the date even
                    // when its ring goes green.
                    drawCircle(
                        color = Color(0xFFD9EDE2).copy(alpha = 0.55f),
                        radius = size.minDimension / 2f,
                        center = Offset(cx, cyTreeMid),
                    )
                    drawCircle(
                        color = SelectionRing.copy(alpha = 0.85f),
                        radius = size.minDimension / 2f,
                        center = Offset(cx, cyTreeMid),
                        style = Stroke(width = 1.5.dp.toPx()),
                    )
                } else if (isToday) {
                    // Today, but not currently the selected day —
                    // user has tapped a past/future day. Coral ring
                    // stays as the "today marker" until they tap
                    // back on today.
                    drawCircle(
                        color = Color(0xFFFCEAE0).copy(alpha = 0.6f),
                        radius = size.minDimension / 2f,
                        center = Offset(cx, cyTreeMid),
                    )
                    drawCircle(
                        color = TodayCoral.copy(alpha = 0.85f),
                        radius = size.minDimension / 2f,
                        center = Offset(cx, cyTreeMid),
                        style = Stroke(width = 1.5.dp.toPx()),
                    )
                }
                drawTree(
                    center  = Offset(cx, cyTreeMid),
                    density = day.density,
                    isToday = isToday,
                )
            }
        }
    }
}

private fun DrawScope.drawTree(center: Offset, density: DayDensity, isToday: Boolean) {
    val cx = center.x
    val cy = center.y
    val canopyTop    = Offset(cx,         cy - 9.dp.toPx())
    val canopyLeft   = Offset(cx - 7.dp.toPx(), cy + 5.dp.toPx())
    val canopyRight  = Offset(cx + 7.dp.toPx(), cy + 5.dp.toPx())
    val canopy = Path().apply {
        moveTo(canopyTop.x, canopyTop.y)
        lineTo(canopyLeft.x, canopyLeft.y)
        lineTo(canopyRight.x, canopyRight.y)
        close()
    }
    val canopyFill = when {
        isToday -> TodayCoral
        else    -> fillFor(density)
    }
    val isHollow = !isToday && density == DayDensity.Empty
    if (isHollow) {
        drawPath(canopy, canopyFill)
        drawPath(
            canopy,
            color = GreenLine.copy(alpha = 0.22f),
            style = Stroke(width = 0.6.dp.toPx()),
        )
    } else {
        drawPath(canopy, canopyFill)
    }
    // Trunk
    val trunkColor = when {
        isToday  -> TodayCoralDark
        isHollow -> TrunkSoft.copy(alpha = 0.4f)
        else     -> TrunkColor
    }
    drawRect(
        color   = trunkColor,
        topLeft = Offset(cx - 1.5.dp.toPx(), cy + 5.dp.toPx()),
        size    = Size(3.dp.toPx(), 4.dp.toPx()),
    )
}

@Composable
private fun LegendRow(
    modifier: Modifier = Modifier,
    onTodayTap: (() -> Unit)? = null,
) {
    Row(
        modifier = modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Text("none", style = AppTypography.Tag, color = AppColors.TextSecondary)
        LegendTree(DayDensity.Empty)
        LegendTree(DayDensity.Light)
        LegendTree(DayDensity.Mid)
        LegendTree(DayDensity.Deep)
        Text("3+ captures", style = AppTypography.Tag, color = AppColors.TextSecondary)
        Spacer(Modifier.weight(1f))
        // Today badge — tappable when [onTodayTap] is wired up; the
        // host pairs the click with snapping the carousel back to
        // today's month and reselecting today's tree.
        val tapMod = if (onTodayTap != null) {
            Modifier
                .clip(RoundedCornerShape(50))
                .clickable(onClick = onTodayTap)
        } else {
            Modifier
        }
        Row(
            modifier = tapMod.padding(horizontal = 4.dp, vertical = 2.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            Box(
                modifier = Modifier
                    .size(20.dp)
                    .clip(RoundedCornerShape(50))
                    .border(1.dp, TodayCoral, RoundedCornerShape(50)),
                contentAlignment = Alignment.Center,
            ) {
                Canvas(Modifier.size(14.dp, 18.dp)) {
                    drawTree(
                        center  = Offset(size.width / 2f, size.height / 2f - 2.dp.toPx()),
                        density = DayDensity.Deep,
                        isToday = true,
                    )
                }
            }
            Text(
                text = "today",
                style = AppTypography.Tag,
                color = if (onTodayTap != null) TodayCoralDark else AppColors.TextSecondary,
            )
        }
    }
}

@Composable
private fun LegendTree(density: DayDensity) {
    Canvas(modifier = Modifier.size(14.dp, 18.dp)) {
        drawTree(
            center  = Offset(size.width / 2f, size.height / 2f - 2.dp.toPx()),
            density = density,
            isToday = false,
        )
    }
}
